crates/chain/src/keychain/txout_index.rs

```diff
@@ -559,32 +559,75 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+) -> SpkIterator {
+    SpkIterator {
+        next_index: range.start,
+        end: range.end,
+        descriptor,
+        secp: Secp256k1::verification_only(),
+    }
+}
+
+#[derive(Clone)]
+pub struct SpkIterator<'a> {
+    next_index: u32,
+    end: u32,
     descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>,
-    range: R,
-) -> impl Iterator<Item = (u32, Script)> + Clone + Send + 'a
-where
-    R: Iterator<Item = u32> + Clone + Send + 'a,
-{
-    let secp = Secp256k1::verification_only();
-    let has_wildcard = descriptor.has_wildcard();
-    range
-        .into_iter()
-        // non-wildcard descriptors can only have one derivation index (0)
-        .take_while(move |&index| has_wildcard || index == 0)
-        // we can only iterate over non-hardened indices
-        .take_while(|&index| index <= BIP32_MAX_INDEX)
-        .map(
-            move |index| -> Result<_, miniscript::descriptor::ConversionError> {
-                Ok((
-                    index,
-                    descriptor
-                        .at_derivation_index(index)
-                        .derived_descriptor(&secp)?
-                        .script_pubkey(),
-                ))
-            },
-        )
-        .take_while(Result::is_ok)
-        .map(Result::unwrap)
+    secp: Secp256k1<bitcoin::secp256k1::VerifyOnly>,
+}
+
+impl<'a> SpkIterator<'a> {
+    // Creates a new script pubkey iterator starting at 0 from a descriptor
+    pub fn new(descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>) -> Self {
+        let secp = Secp256k1::verification_only();
+        let end = if descriptor.has_wildcard() {
+            BIP32_MAX_INDEX
+        } else {
+            0
+        };
+
+        Self {
+            next_index: 0,
+            end,
+            descriptor,
+            secp,
+        }
+    }
+}
+
+impl<'a> Iterator for SpkIterator<'a> {
+    type Item = (u32, Script);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // for non-wildcard descriptors:
+        // * we expect the first element to be Some((0, spk)), then will return None after.
+        // for wildcard descriptors:
+        // * we expect it to keep iterating until exhausted.
+        let has_wildcard = self.descriptor.has_wildcard();
+        if self.next_index >= self.end || (has_wildcard && self.next_index >= BIP32_MAX_INDEX) {
+            return None;
+        }
+
+        let script = self
+            .descriptor
+            .at_derivation_index(self.next_index)
+            .derived_descriptor(&self.secp)
+            .expect("the descriptor cannot need hardened derivation")
+            .script_pubkey();
+        let output = (self.next_index, script);
+
+        self.next_index += 1;
+
+        Some(output)
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        //let n = self.next_index + n as u32;
+        //self.descriptor.at_derivation_index(n);
+        self.next_index = self.next_index.saturating_add(n as u32);
```

Comment:
`n` (which is `usize`) may be >32bit (depending on the architecture). We should use a fallible method `TryInto::try_into` and use `u32::MAX` for the error case.